lib/AST/ASTPrinter.cpp

```diff
@@ -3749,19 +3757,37 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
   }
 
   void visitParentType(Type T) {
+    /// Don't print the parent type if it's being printed in that type context.
+    if (Options.TransformContext) {
+      if (auto currentType = Options.TransformContext->getBaseType()) {
+        auto printingType = T;
+        if (currentType->hasArchetype())
+          currentType = currentType->mapTypeOutOfContext();
+
+        if (auto errorTy = printingType->getAs<ErrorType>())
+          if (auto origTy = errorTy->getOriginalType())
+            printingType = origTy;
+
+        if (T->hasArchetype())
```

Comment:
This is a bit weird, shouldn't it be 'if (printingType->hasArchetype())'?

Also do you need to map both currentType and printingType out of context? It should just work to compare them directly.